Use PATCH to resume/pause deployment

mfojtik opened the pull request:
Alternative version of: https://github.com/kubernetes/kubernetes/pull/30919
Fixes: https://github.com/kubernetes/kubernetes/issues/20437

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31220)

<!-- Reviewable:end -->

mfojtik commented:
@smarterclayton can you fix me? ;-) ^^

mfojtik commented:
@fraenkel PTAL

mfojtik commented:
@smarterclayton btw. https://github.com/kubernetes/kubernetes/pull/31234 just got `cla: yes`

kargakis commented:
@k8s-bot test this issue: #IGNORE

kargakis commented:
@kubernetes/kubectl

kargakis commented:
@k8s-bot test this issue: #IGNORE

kargakis commented:
@mfojtik please have a look at the test failure

mfojtik commented:
@kargakis it seems none of them are caused by this patch unless I'm missing something

krmayankk commented on pkg/client/unversioned/deployment.go:
```diff
@@ -82,6 +83,13 @@ func (c *deployments) Create(deployment *extensions.Deployment) (result *extensi
+	err = c.client.Patch(api.StrategicMergePatchType).Namespace(c.ns).Resource("deployments").SubResource(subresources...).Name(name).Body(data).Do().Into(result)
+	return
+}
+
```
what is the difference between calling Update any time many k8s yaml with full yaml contents changes vs just calling  Patch. Is it that Patch can take a diff and apply it to an existing k8s deployment ?

kargakis replied:
Yes. Update fails on update conflicts whereas Patch doesn't. Since the pause field is supposed to be manipulated only via `pause` and `resume` (and `delete`) then it makes sense to patch in these commands.

mfojtik replied:
the `PATCH` endpoints should be used only from CLI tools (not by controllers or internally).

kargakis commented:
@mfojtik can you rebase this?

mfojtik commented:
yes as a first thing tomorrow

0xmichalis reviewed (changes requested):
rebase

kargakis commented:
It's still unclear to me why I haven't seen integration tests green. @mfojtik do they pass for you locally?

mfojtik commented:
@kargakis not yet, but will try

kargakis commented:
@mfojtik can you rebase

kargakis commented:
@k8s-bot test this issue: #IGNORE

kargakis commented:
```
FAIL!
Get deployment {{range.items}}{{.spec.paused}}:{{end}}
  Expected: true:true:
  Got:      <no value>:<no value>:
(B[m[31m
1451 hack/make-rules/test-cmd.sh
```

mfojtik commented:
@kargakis will check

kargakis commented:
```
message:error: unable to decode "hack/testdata/recursive/rc/rc/busybox-broken.yaml": Object 'Kind' is missing in '{"apiVersion":"v1","ind":"ReplicationController","metadata":{"labels":{"app":"busybox2"},"name":"busybox2"},"spec":{"replicas":1,"selector":{"app":"busybox2"},"template":{"metadata":{"labels":{"app":"busybox2"},"name":"busybox2"},"spec":{"containers":[{"command":["sleep","3600"],"image":"busybox","imagePullPolicy":"IfNotPresent","name":"busybox"}],"restartPolicy":"Always"}}}}'
has not:error when pausing "hack/testdata/recursive/rc/busybox.yaml
```
I think this is happening because Patch will succeed even with invalid YAML?

AdoHe commented:
@kargakis patch content is in yaml format?

kargakis commented:
> @kargakis patch content is in yaml format?

No, JSON. I was talking about the yaml fixture that triggers the test to fail.

mfojtik commented:
@kargakis I think it just returns silently... i need to check how the errors are reporting when you supply an invalid file.

mfojtik commented:
@kargakis that was a good error catch by tests :-) what happened was that when the `mutateFn` in `CalculatePatch` failed and returned "no-ok", we swallowed the error and moved on. I fixed that so we return the patch (which is empty, but carries the error).

mfojtik commented on pkg/kubectl/cmd/set/helper.go:
```diff
@@ -120,9 +120,12 @@ func CalculatePatches(infos []*resource.Info, encoder runtime.Encoder, mutateFn
 
 		ok, err := mutateFn(info)
-		if !ok {
+		if !ok && err == nil {
```
@kargakis talking about this (the fix above is also needed as we swallowing the encoding errors (the `patch.Err` is overrided by the code below.... ))

mfojtik replied:
actually I think that maybe returning "ok" from the mutateFn in case of error might be better... @kargakis @deads2k what you think?

kargakis commented:
needs a rebase

mfojtik committed:
Use PATCH when pausing/resuming objects and CalculatePatches to get the patch

mfojtik commented:
@kargakis rebased, but need review of the factory interface function names.

kargakis commented:
New factory changes lgtm

mfojtik commented:
@kargakis test fixed, ready for shipping.

kargakis commented:
/lgtm

kargakis commented:
:tada:
